backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/IssuerManager.java

```diff
@@ -129,4 +162,139 @@ public CredDef createCredDef(@NonNull String schemaId, @NonNull String tag, bool
+        }
+        final Optional<BPACredentialDefinition> dbCredDef = credDefRepo.findById(credDefId);
+        if (!dbCredDef.isPresent()) {
+            throw new IssuerException(String.format("Could not find credential definition with id '%s'", credDefId));
+        }
+        try {
+            Optional<V1CredentialExchange> exchange = ac.issueCredentialSend(
+                    V1CredentialProposalRequest
+                            .builder()
+                            .connectionId(dbPartner.get().getConnectionId())
+                            .schemaId(dbCredDef.get().getSchema().getSchemaId())
+                            .credentialProposal(
+                                    new CredentialPreview(
+                                            CredentialAttributes.from(document)))
+                            .credentialDefinitionId(dbCredDef.get().getCredentialDefinitionId())
+                            .build());
+            if (exchange.isPresent()) {
+                // should this go in a handler for offer_sent?
+                HashMap<String, Object> cp = new Gson().fromJson(exchange.get().getCredentialProposalDict(),
+                        HashMap.class);
+                HashMap<String, Object> co = new Gson().fromJson(exchange.get().getCredentialOfferDict(),
+                        HashMap.class);
+                // NOTE: using toLowerCase, ColumnTransformer not working :(
+                BPACredentialExchange cex = BPACredentialExchange.builder()
+                        .type(CredentialType.SCHEMA_BASED)
+                        .schema(dbCredDef.get().getSchema())
+                        .partner(dbPartner.get())
+                        .credDef(dbCredDef.get())
+                        .role(exchange.get().getRole().name().toLowerCase(Locale.getDefault()))
+                        .state(exchange.get().getState().name().toLowerCase(Locale.getDefault()))
+                        .threadId(exchange.get().getThreadId())
+                        .credentialExchangeId(exchange.get().getCredentialExchangeId())
+                        .credentialProposal(cp)
+                        .credentialOffer(co)
+                        .updatedAt(TimeUtil.parseZonedTimestamp(exchange.get().getUpdatedAt()))
+                        .build();
+
+                BPACredentialExchange saved = credExRepo.save(cex);
+                return CredEx.from(saved);
+            } else {
+                throw new IssuerException(String.format("Could not issue the credential for definition '%s'",
+                        dbCredDef.get().getCredentialDefinitionId()));
+            }
+        } catch (IOException e) {
+            throw new NetworkException("No aries connection", e);
+        }
+    }
+
+    public List<CredEx> listCredentialExchanges(String role) {
+        List<BPACredentialExchange> exchanges = new ArrayList<>();
```

Comment:
Does the role filter work? Basically we now have two worlds: 1 credentials that have been issued to the BPA that are stored in MyCredential and 2 credentials that have been issued by the BPA which are stored in the exchange log. Which sounds like we should rethink the whole concept a bit ;)

Reply:
not sure if we will have a case to store any holder information in this table, but leaving for now. it should become clearer as we turn off auto flags.